Cancel truncation of zero-extensions in simplifier

jdmpapin opened the pull request:
Simplifier already cancels truncation of sign-extensions, and also of
zero-extensions in some cases. It doesn't matter that the high bits are
the result of zero-extension, since the high bits are discarded anyway.

The following now cancel:
* `i2s` of `su2i`
* `i2b` of `bu2i`
* `s2b` of `bu2s`

vijaysun-omr commented:
@jdmpapin Are there any other very similar conversions patterns involving unsigned variants that could also be caught ?

jdmpapin commented:
Among our 64-, 32-, 16-, and 8-bit integer types, the truncation operations are `l2i`, `l2s`, `l2b`, `i2s`, `i2b`, and `s2b`.

`l2i`, `l2s`, and `l2b` use `intDemoteSimplifier`, which already cancels both zero-extensions and sign-extensions, as long as the comments on `inverseOp1` and `inverseOp2` are to be believed.

`i2s` is already dealt with in this change.

`i2b` looks like it used to cancel `bu2i`, but the logic to do so was commented out for some reason.

`s2b` could stand to start cancelling `bu2s`.

vijaysun-omr commented:
I would like @andrewcraik to review this. I would like us to also consider fixing the couple of cases Devin mentioned in the previous comment, namely 

"i2b looks like it used to cancel bu2i, but the logic to do so was commented out for some reason.
 s2b could stand to start cancelling bu2s."

andrewcraik commented:
This change looks good to me. Looking at the i2b and bu2i cancellation there is no good reason for that to be disabled - the disable does not look to have been done for the right reasons - the cancellation is valid and safe. If the s2b and bu2s cancellation is another two-liner I think we should put all 3 in together since it make the simplifier better overall. @jdmpapin let us know once the change is updated to handle all 3 cancellation cases and we can look to get this merged.

jdmpapin commented:
I've updated the change to cancel all three of the truncation/zero-extension pairs that weren't already cancelling, and I updated the commit message and PR description to match.

andrewcraik commented:
change looks good to me - thank you for revision. @vijaysun-omr can you please review and merge now that your concerns have been addressed?

vijaysun-omr commented:
@jdmpapin Thank you for making the broader change. Merging now.
